include/arch/arm/arch.h

```diff
@@ -306,55 +354,171 @@ typedef u32_t k_mem_partition_attr_t;
 
 #ifdef CONFIG_ARM_USERSPACE
 #ifndef _ASMLANGUAGE
+
 /* Syscall invocation macros. arm-specific machine constraints used to ensure
- * args land in the proper registers. Currently, they are all stub functions
- * just for enabling CONFIG_USERSPACE on arm w/o errors.
+ * args land in the proper registers.
  */
+extern int _arm_do_syscall(u32_t call_id, u32_t arg1,
+	u32_t arg2, u32_t arg3, u32_t arg4, u32_t arg5, u32_t arg6);
 
 static inline u32_t _arch_syscall_invoke6(u32_t arg1, u32_t arg2, u32_t arg3,
 					  u32_t arg4, u32_t arg5, u32_t arg6,
 					  u32_t call_id)
 {
-	return 0;
+	register u32_t ret __asm__("r0") = arg1;
+	register u32_t r1 __asm__("r1") = arg2;
+	register u32_t r2 __asm__("r2") = arg3;
+	register u32_t r3 __asm__("r3") = arg4;
+
+	__asm__ volatile("sub sp, #16\n"
+			 "str %[a5], [sp, #0]\n"
+			 "str %[a6], [sp, #4]\n"
+			 "str %[cid], [sp, #8]\n"
+			 "svc %[svid]\n"
+			 "add sp, #16\n"
+			 : "=r"(ret)
+			 : [cid] "r" (call_id),
+			   [svid] "i" (_SVC_CALL_SYSTEM_CALL),
+			   "r" (ret), "r" (r1), "r" (r2), "r" (r3),
+			   [a5] "r" (arg5), [a6] "r" (arg6)
+			 : "ip", "memory");
+
+	return ret;
 }
 
 static inline u32_t _arch_syscall_invoke5(u32_t arg1, u32_t arg2, u32_t arg3,
 					  u32_t arg4, u32_t arg5, u32_t call_id)
 {
-	return 0;
+	register u32_t ret __asm__("r0") = arg1;
```

Comment:
are these actually necessary?
does it do the same thing if you just put the arguments directly in the __asm__ call? the "r" constraint should ensure that they end up in registers

Reply:
disregard, we need these to end up in *specific* registers and ARM doesn't have constraint codes for them